### PR TITLE
MCO-1152: MCO-1146: Add e2e tests for NodeDisruptionPolicy

### PR DESCRIPTION
**Pull request by djoshy**
This PR adds:
- some e2e tests for NodeDisruptionPolicy which landed in #4267 
- add a generation check before the MCD consumes NodeDisruptionPolicy Status
- the MCD will now degrade on NodeDisruptionPolicy errors

### CODE REVIEW

File: test/e2e-techpreview/nodedisrupt_test.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+	testFuncs := []func(*testing.T, corev1.Node, []opv1.NodeDisruptionPolicySpecAction){testFilePolicy, testUnitPolicy, testSSHKeyPolicy}
+
+	for testIndex, testFunc := range testFuncs {
+		testIndex := testIndex
+		testFunc := testFunc
+		t.Run(helpers.GetFunctionName(testFunc), func(t *testing.T) {
+			// Only parallelize if there are enough nodes to run the tests individually
+			if len(nodes) >= len(testFuncs) {
+				t.Parallel()
```
Comment (cheesesashimi) [inline test/e2e-techpreview/nodedisrupt_test.go:69]: praise: Nice! I want to be able to parallelize more in our e2e suite.

File: pkg/daemon/update.go [hunk 2]
```diff
@@ -736,13 +736,11 @@ func (dn *Daemon) calculatePostConfigChangeNodeDisruptionAction(diff *machineCon
...
-		// TODO: When Conditions on this object are implemented; this check could be updated to only proceed when
-		// status.ObservedGeneration matches the last generation of MachineConfiguration
-		if len(mcop.Status.NodeDisruptionPolicyStatus.ClusterPolicies.Files) == 0 {
-			klog.Errorf("calculating NodeDisruptionPolicies: NodeDisruptionPolicyStatus has not been populated yet")
-			err = fmt.Errorf("NodeDisruptionPolicyStatus has not been populated yet")
+
+		// Ensure status.ObservedGeneration matches the last generation of MachineConfiguration
+		if mcop.Generation != mcop.Status.ObservedGeneration {
+			klog.Errorf("calculating NodeDisruptionPolicies: NodeDisruptionPolicyStatus is not up to date.")
+			err = fmt.Errorf("NodeDisruptionPolicyStatus is not up to date")
```
[the remark this answers is not in the document]
Comment (cheesesashimi) [reply pkg/daemon/update.go:743]: thought: I just took a look at why this was written like this. Basically, we use the variable `err` for _both_ the value returned by `wait.PollUntilContextTimeout()` in addition to what is inside the closure that `wait.PollUntilContextTimeout()` executes. Its a bit confusing as to whether we should stop polling in that case or keep going and deal with the error later. If we want to keep going and deal with the error later, using the [Aggregate](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/errors/errors.go#L46) coupled with an appropriate deduplication function could help readability.

(To be clear: I am not asking you to change this as part of this PR. It's just a thought I'm putting here for posterity.)

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
Overall this looks great! I just have a few small suggestions.

File: test/e2e-techpreview/nodedisrupt_test.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	cs := framework.NewClientSet("")
+	t.Cleanup(helpers.CreatePoolAndApplyMCToNode(t, cs, testMCPFileName, node, nil))
+
+	// Step through each action and check if it was applied correctly. This loop generates a MachineConfiguration
+	// and a MachineConfig for each testAction and verifies that the correct NodeDisruptionPolicy was applied.
+	for _, action := range testActions {
+		var ac *mcoac.NodeDisruptionPolicySpecFileApplyConfiguration
+		fileName := "/etc/test-" + string(action.Type)
+		if action.Type == opv1.ReloadSpecAction {
+			ac = mcoac.NodeDisruptionPolicySpecFile().WithPath(fileName).WithActions(mcoac.NodeDisruptionPolicySpecAction().WithType(action.Type).WithReload(mcoac.ReloadService().WithServiceName(action.Reload.ServiceName)))
```
[the remark this answers is not in the document]
Comment (djoshy) [reply test/e2e-techpreview/nodedisrupt_test.go:91]: Will attempt to clean this up, thanks (:

File: test/helpers/utils.go [hunk 4]
```diff
@@ -760,6 +762,17 @@ func AssertFileNotOnNode(t *testing.T, cs *framework.ClientSet, node corev1.Node
 		assert.Contains(t, out, "No such file or directory", "expected command output to contain 'No such file or directory', got: %s", out)
 }
 
+// Asserts that a given file exists on the underlying node and returns the content of the file.
```
[the remark this answers is not in the document]
Comment (djoshy) [reply test/helpers/utils.go:765]: I believe we have helpers that checks if a file exists/does not exist, but not one that grabs the content in the file :)

File: test/e2e-techpreview/nodedisrupt_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+// place when registried are added. It also verifies that crio reloads takes place in both cases.
+func TestNodeDisruptionPolicySpecialAction(t *testing.T) {
+
+	t.Logf("Verifing Special action")
+
+	cs := framework.NewClientSet("")
+	t.Cleanup(helpers.CreatePoolAndApplyMC(t, cs, testMCPSpecialName, nil))
+	nodeUnderTest := helpers.GetSingleNodeByRole(t, cs, testMCPSpecialName)
+
+	machineConfigClient := machineconfigclientset.NewForConfigOrDie(cs.GetRestConfig())
```
[the remark this answers is not in the document]
Comment (djoshy) [reply test/e2e-techpreview/nodedisrupt_test.go:162]: Oh was not aware of that - I'll try to use this instead! Thanks!

File: pkg/daemon/update.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (djoshy) [reply pkg/daemon/update.go:743]: Sorry about the structure being confusing! My thought process to use it was the following:
- If we encounter an error, we don't want to keep going in the polling function. We want to try again after an interval of time.
- If we timeout on the polling function, we want the last error we encountered to be reported. Hence the use of the same variable inside and outside of the function.

I hope that clears up why I went with the last error vs aggregation. Any of the errors within the polling loop are fatal, with the earlier ones being slightly more fatal than the ones following.

File: test/e2e-techpreview/nodedisrupt_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djoshy) [reply test/e2e-techpreview/nodedisrupt_test.go]: I was able to break this down as suggestion, and create a new helper - making it overall a lot cleaner and re-usable. Thanks so much for this suggestion!

File: pkg/daemon/update.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cheesesashimi) [reply pkg/daemon/update.go:743]: Ahhh that makes sense, thanks for the clarification!

[the remark this answers is not in the document]
Comment (djoshy) [reply pkg/daemon/update.go]: I assigned a new variable so it is easier to read what I'm trying to do (:

**Review by sinnykumari — COMMENTED** [review]
changes looks good.
/approve
Will see if Zack have any final thoughts and would like to tag lgtm.

**Comment by djoshy** [comment]
/hold

just in case QE wants to take a look

**Comment by djoshy** [comment]
Discussed with @sergiordlr , as this is mainly e2es, pre merge QE is not required. Unholding.

/unhold
/retest-required

**Pull request merged**